public/static/js/init.py

```diff
@@ -799,29 +811,48 @@ def run_turtle_cmd(msg):
 def hit_breakpoint(lineno, alocals, aglobals, is_secondary=False):
     global step_into
     global last_seen_lineno
+    global watches
+    global in_watch
     if is_secondary and last_seen_lineno == lineno:
         return True
     last_seen_lineno = lineno
-    if step_into or lineno in active_breakpoints:
+    if not in_watch and (step_into or lineno in active_breakpoints):
         step_into = False
         # remove wrapper and breakpt method
-        stack = traceback.extract_stack()[1:-1]
+        #stack = traceback.extract_stack()[1:-1]
         VARS_TO_REMOVE = ["__name__", "__main__", "__package__", "__annotations__", "__doc__",
                           "__loader__", "__spec__", "__builtins__", "sys", "js", "ast", "MyOutput", "my_output",
                           "pydebug", "input", "hit_breakpoint", "VARS_TO_REMOVE", "traceback", "sleep", "os", "time", "last_seen_lineno"]
-        globals = {k: repr(
-            aglobals[k]) for k in aglobals if k not in VARS_TO_REMOVE and not callable(aglobals[k])}
-        if alocals != aglobals:
-            locals = {k: repr(
-                alocals[k]) for k in alocals if k not in VARS_TO_REMOVE and not callable(alocals[k])}
-        else:
-            locals = {}
-
-        post_message({"cmd": "breakpt", "lineno": lineno,
-                     "globals": globals, "locals": locals})
-        resp = json.loads(synchronise('/@debug@/break.js'))
-        if (resp.get("breakpoints")):
-            update_breakpoints(resp["breakpoints"])
-        if (resp.get("step")):
-            step_into = True
+        stay = True
+        while stay:
```

Comment:
"stay": true flag in the JS response allows the client to request the updated value of the watches without proceeding to the next line